File: OJ/Leetcode/Algorithm/404. Sum of Left Leaves.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sumOfLeftLeaves(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        return self.dfs(root, False)
    def dfs(self, root, flag):
        if not root:
            return 0
        if not root.left and not root.right and flag:
            return root.val
        ans = self.dfs(root.left, True)
        ans += self.dfs(root.right, False)
        return ans
        

File: OJ/Leetcode/Algorithm/404. Sum of Left Leaves.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int ret = 0;
    int sumOfLeftLeaves(TreeNode* root) {
        if (!root) return 0;
        if (!root->left && !root->right) return 0;
        dfs(root, 0);
        return ret;
    }
    void dfs(TreeNode* root, int a) {
        if (!root) return ;
        if (!root->left && !root->right && a== 0) ret += root->val;
        else if (!root->left && root->right) dfs(root->right,1);
        else if (!root->right && root->left) dfs(root->left, 0);
        else {
            dfs(root->right,1);
            dfs(root->left, 0);
        }
    }
};
